torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -566,7 +574,11 @@ void ProcessGroupNCCL::workCleanupLoop() {
          /* no increment*/) {
       auto& work = *it;
       if (work->isCompleted()) {
-        // Remove all Completed WorkNCCL Objects from the Vector
+        // Handle Exceptions on failed GPU operations and remove completed
+        // workNCCL objects from work vector.
+        if (work->finishedGPUExecution()) {
+          work->handleNCCLGuard();
```

Comment:
Throwing an exception here is going to kill this thread and also cause a process crash. This doesn't seem like a good idea in general. What is the reason to throw the exception from this thread itself?

Reply:
Referring to Hongyi's comment (https://github.com/pytorch/pytorch/pull/43232#discussion_r482569599). The premise of this work is prevent stuck/hanging GPU ops from from stalling the entire training process by throwing an exception (crashing the process), and then allowing something like torchelastic to recover and continue training from last checkpoint. We need a thread that periodically checks for exceptions (either due to errors or timeouts) in unfinished collectives and throws these exceptions, which we do in the workCleanupThread.

Reply:
In case of errors or timeouts we actually abort the NCCL communicators. In that case subsequent NCCL calls will fail and an error would be raised to the user. Is this sufficient for what we're looking to do here?

My concern with this approach is that it drastically changes the default behavior for users today and many users don't actually use torchelastic. It would be pretty bad user experience if by default the process crashes on any NCCL errors/timeouts.